Source/MLX/MLXArray.swift

```diff
@@ -348,6 +351,50 @@ public final class MLXArray {
         }
     }
 
+    /// Return the contents as a Metal buffer in the native ``dtype``.
+    ///
+    /// > If you can guarantee the lifetime of the ``MLXArray`` will exceed the MTLBuffer and that
+    /// the array will not be mutated (e.g. using indexing or other means) it is possible to pass `noCopy: true`
+    /// to reference the backing bytes.
+    ///
+    /// ### See Also
+    /// - <doc:conversion>
+    /// - ``asArray(_:)``
+    /// - ``asData(noCopy:)``
+    public func asMTLBuffer(device: any MTLDevice, noCopy: Bool = false) -> (any MTLBuffer)? {
```

Comment:
Could this not be implemented as a simple wrapper on top of `asData`?

Reply:
I am not so sure about making a MTLBuffer without copying the bytes: you need to be very careful with the lifetime and I am not sure there is anything you can do with it synchronously where you can be sure of this, but the comment says the same thing as the `asData` does -- it is up to the caller to make sure of that if they go that route.

Reply:
Let's wait until I have the contiguous backing change landed -- I think that will simplify this a bit.  Here is what I am thinking for the body of asData():

```swift
        if noCopy && self.contiguousToDimension() == 0 {
            // the backing is contiguous, we can provide a wrapper
            // for the contents without a copy (if requested)
            let source = UnsafeMutableRawPointer(mutating: mlx_array_data_uint8(self.ctx))!
            return Data(
                bytesNoCopy: source, count: self.nbytes,
                deallocator: .none)
        } else {
            let source = UnsafeRawBufferPointer(start: mlx_array_data_uint8(self.ctx), count: physicalSize * itemSize)

            var data = Data(count: self.nbytes)
            data.withUnsafeMutableBytes { destination in
                copy(from: source, to: destination)
            }
            return data
        }
```

This MTLBuffer code could be adapted to use something like that.

Reply:
ok #96 is in -- take a look at `asData`:

https://github.com/ml-explore/mlx-swift/blob/main/Source/MLX/MLXArray.swift#L379

I am not sure the `noCopy` is really usable -- it works for Data because you might copy it somewhere else or write it in a file and you can do all of that synchronously.  For a metal buffer, there isn't _anything_ you can do synchronously, so keeping the pointer to the interior probably isn't safe.  I think you could just use the `else` case where it copies it.